### PR TITLE
CA-286291: Fix the crash when quickly and repeatedly popping up Migrate-to-Server menu

### PR DESCRIPTION
**Pull request by michael2012z**
The PR aims to fix the crash when a user pops up Migrate-to-Server menu repeatedly and quickly. The problem is more probable to happen if the server list is long.
The root cause is two or more threads can be created to update the menu in parallel, if the user operates quickly enough, which definitely lead to a program crash.
The solution is to guarantee only one thread is working on menu update. If the previous menu has been closed and a new menu is to be made, we need to make sure the previous job quit safely (don't invoke main window any more) before starting a new job.

### CODE REVIEW

File: XenAdmin/Commands/Controls/VMOperationToolStripMenuItem.cs [hunk 1]
```diff
@@ -125,205 +124,279 @@ protected override void OnDropDownOpening(EventArgs e)
...
+
+        private class HostListUpdater
         {
-            SelectedItemCollection selection = Command.GetSelection();
-            // set the first menu item to be the WLB optimal server menu item
-            VMOperationToolStripMenuSubItem firstItem = (VMOperationToolStripMenuSubItem)base.DropDownItems[0];
-            var firstItemCmd = new VMOperationWlbOptimalServerCommand(Command.MainWindowCommandInterface, selection, _operation, recommendations);
-            var firstItemCmdCanExecute = firstItemCmd.CanExecute();
+            private ProduceConsumerQueue workQueue = new ProduceConsumerQueue(1);
+            private ProduceConsumerQueue workerQueueWithouWlb = new ProduceConsumerQueue(25);
```
Comment (MihaelaStoica) [inline XenAdmin/Commands/Controls/VMOperationToolStripMenuItem.cs]: I believe this should have been called workerQueueWithoutWlb (missing 't')

Comment (MihaelaStoica) [reply XenAdmin/Commands/Controls/VMOperationToolStripMenuItem.cs]: workerQueueWithouWlb is always recreated in updateHostList, so I don't think it is necessary to be created here as well, especially as the ProduceConsumerQueue constructor creates and starts 25 threads.

File: XenAdmin/Commands/Controls/VMOperationToolStripMenuItem.cs [hunk 2]
```diff
@@ -125,205 +124,279 @@ protected override void OnDropDownOpening(EventArgs e)
...
-                    if (recommendations.IsError)
-                        EnableAppropriateHostsNoWlb(session);
-                    else
-                        EnableAppropriateHostsWlb(session, recommendations);
-                }
-                else
-                {
-                    EnableAppropriateHostsNoWlb(session);
-                }
+                hostListUpdater.updateHostList(this, session);
```
[the remark this answers is not in the document]
Comment (michael2012z) [reply XenAdmin/Commands/Controls/VMOperationToolStripMenuItem.cs]: I used too much threads here, it was a problem in this solution.
I will update the thread model.

File: XenAdmin/Commands/Controls/VMOperationToolStripMenuItem.cs [hunk 3]
```diff
@@ -120,210 +116,277 @@ protected override void OnDropDownOpening(EventArgs e)
...
             }
-            
-            // Shuffle the list to make it look cool
-            // Helpers.ShuffleList(hostMenuItems);
 
-            // sort the hostMenuItems by star rating
-            hostMenuItems.Sort(new WlbHostStarCompare());
+            public void Stop()
+            {
+                _stopped = true;
```
[the remark this answers is not in the document]
Comment (michael2012z) [reply XenAdmin/Commands/Controls/VMOperationToolStripMenuItem.cs]: Sorry, what I wanted to write is "Stopped = true;" In fact this is the only case we use the setter of Stopped field.

**Comment by kc284** [comment]
Could we please rebase this PR on master?

**Comment by MihaelaStoica** [comment]
I am merging thi PR now and I'll submit the changes requested by @kc284 on a separate PR.

**Pull request merged**